src/serde_helpers.rs

```diff
@@ -182,6 +238,8 @@ macro_rules! impl_deserialize_for_internally_tagged_enum {
                             $(
                                 $variant_tag => Ok(deserialize_variant!( de, $enum, $($variant)+ )),
                             )*
+                            $( _ => Ok(deserialize_variant!( de, $enum, $($default_variant)+ )), )?
+                            #[allow(unreachable)]
```

Comment:
You can use intermediate macro with two possible variants to select required behavior. General idea:
```rust
macro_rules! def {
  // variant 1
  (
    $($variant_tag:literal),*
    _ => Ok( $tokens:tt )
  ) => {
    _ => Ok($tokens),
  },
  // variant 2
  (
    $($variant_tag:literal),*
  ) => {
    _ => Err(A::Error::unknown_field(&tag, &[$($variant_tag),+])),
  },
}

// use 
def!(
  $($variant_tag),*
  $( _ => Ok(deserialize_variant!( de, $enum, $($default_variant)? )), )?
)
```
If `$(...)?` part exists, then the first alternative would be selected, otherwise the second. That way we would not need to use `#[allow(unreachable)]` attribute, which could be forbidden in the project via `#[forbid(unreachable)]`

Reply:
Thanks for the pointer, I was had the conviction the allow_unreachable wasn't a clean solution.

I get the general idea here, but this code doesn't compile ("no rules expected the token `,`") and I have no clue what I should change. 

Thanks.